joomla/joomla-cms
# Formfield layouts

### Summary of Changes

This is a resubmission of https://github.com/joomla/joomla-cms/pull/13018 because I totally abandoned @Spudley which is the worst I could do with a new contributor :( I'm really sorry mate and I hope you get your contribution merged at least 2 years later. I created a custom email rule for the repo notifications and everything is messed there.

This pull request adds the default layout paths to `FormField::getLayoutPaths()` so developers can inherit them and add any custom layout path required by their extensions. 

Before this patch a developer wanting to set custom layout paths had to do something like:

```php
	protected function getLayoutPaths()
	{
		$renderer = new FileLayout('default');

		return array_merge(
		    array(__DIR__ . '/custom-dir'), 
		    $renderer->getDefaultIncludePaths())
	}
```

After this patch it should be easier like:

```php
	protected function getLayoutPaths()
	{
		return array_merge(
			[__DIR__ . '/custom-dir'],
			parent::getLayoutPaths()
		);
	}
```

which should be better because:

* Parent cares about default include paths.
* Child classes do not have to explicitly create an instance of `FileLayout`

This also adds unit tests for `FormField::getLayoutPaths()`.

### Testing Instructions

Edit any form containing fields (for instance com_content article) and ensure that form fields are shown correctly.

If you want ot test layout paths customisation you can create a field like it's done in the tests:

https://github.com/phproberto/joomla-cms/blob/formfield-layouts/tests/unit/suites/libraries/joomla/form/_testfields/customlayouts.php
